### PR TITLE
Polygon holes

### LINKED ISSUE #1729: add support for multipolygons and polygon interior holes
Punted from #1655. Will need to be exposed in `MGLPolygon` as well on iOS.

### PR DESCRIPTION
**Pull request by 1ec5**
A slapdash attempt at reviving #5170 to add support for holes in `Polygon` and `PolygonOptions`. I resolved the conflicts without the aid of an IDE, so this PR needs close scrutiny – there’s probably quite a bit of copy 🍝 in here.

Fixes #1729.

### CODE REVIEW

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Polygon.java [hunk 1]
```diff
@@ -2,18 +2,48 @@
...
 
+import java.util.ArrayList;
+import java.util.List;
+
 /**
  * Polygon is a geometry annotation that's a closed loop of coordinates.
  */
 public final class Polygon extends MultiPoint {
 
+    private List<List<LatLng>> holes;
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Polygon.java]: Could you make this a `final` list and initialize in declaration?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Polygon.java [hunk 2]
```diff
@@ -2,18 +2,48 @@
...
+    }
+
+    /**
+     * Sets the holes of this polygon. This method will take a copy of the holes, so further
+     * mutations to holes will have no effect on this polygon.
+     *
+     * @param points A {@link List} {@link List}s of {@link LatLng} points making up the holes.
+     */
+    public void setHoles(List<? extends List<LatLng>> holes) {
+        this.holes = new ArrayList<>(holes);
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Polygon.java]: Can just update the member list with the collection, so the reference is not exposed.

File: platform/android/src/jni.cpp [hunk 3]
```diff
@@ -757,7 +758,26 @@ jni::jarray<jlong>* nativeAddPolygons(JNIEnv *env, jni::jobject* obj, jlong nati
...
+
+        jni::jobject* holes = jni::GetField<jni::jobject*>(*env, polygon, *polygonHolesId);
+        NullCheck(*env, holes);
+        jni::jarray<jni::jobject>* jarrayHoles =
+            reinterpret_cast<jni::jarray<jni::jobject>*>(jni::CallMethod<jni::jobject*>(*env, holes, *listToArrayId));
+        NullCheck(*env, jarrayHoles);
+
+        std::size_t size = jni::GetArrayLength(*env, *jarrayHoles);
+        for (std::size_t j = 0; j < size; j++) {
+            jni::jobject* hole = reinterpret_cast<jni::jobject*>(jni::GetObjectArrayElement(*env, *jarrayHoles, j));
```
Comment (ivovandongen) [inline platform/android/src/jni.cpp]: This cast is redundant, `jni::GetObjectArrayElement` returns `jni::jobject*` already

**Comment by 1ec5** [comment]
Thanks for taking a look!

> I don't see any holes in the example.

I think the example originally worked, but now the star's outer ring has been moved to a static member of the class instead of a local array. For the purposes of this example, I think we'll be able to define the hole similarly to the outer ring instead of trying to define the hole dynamically.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Polygon.java (hunk 2 above)
Comment (1ec5) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Polygon.java]: You mean `this.holes = holes`? I think the idea was to make a copy of the list, similar to `MultiPoint.setPoints()`.

File: platform/android/src/jni.cpp (hunk 3 above)
Comment (1ec5) [inline platform/android/src/jni.cpp]: Fixed in 2ec060543899367c3b1f9026117488d8c133ec10.

**Comment by 1ec5** [comment]
I modernized the example activity, though I haven’t tried it yet.

**Comment by zugaldia** [comment]
@1ec5 Polygon looks different, but I still see no holes:

screenshot_20170202-150032

**Comment by 1ec5** [comment]
Thanks @zugaldia. I need to actually get around to setting up a development environment. GitHub comments don't make for a good debugger. 😅

**Comment by zugaldia** [comment]
@1ec5 totally :) wanna work on that today to benefit from the small delta in our coordinates?

**Comment by caHarkness** [comment]
I see that there's a pull request open:
`add support for multipolygons and polygon interior holes #1729`

Does this mean that in a future version, it is guaranteed that we will have interior holes on Android?  I've been following this issue for a while and have seen iOS get a lot of attention on the subject, but I am more interested in Android.

**Comment by 1ec5** [comment]
#1729 is the bug report; this is the PR that would fix it for Android. The issue was fixed for iOS and macOS in #5110, which also involved adding support in the core C++ code for polygon holes. This PR plumbs that support through JNI to the Android SDK and adds an activity to demonstrate the feature. Unfortunately, the activity currently demonstrates that the JNI implementation is broken. @Guardiola31337 is currently assigned to take a look. We’re still scoping out what the next release will contain, but this PR is on our radar.

**Comment by 1ec5** [comment]
#8557 landed instead.

**Pull request closed without merging**